### ISSUE TITLE
feat: allow to declare development environment

### ISSUE THREAD

**Issue by romm** [CuyZ/Valinor#24]
Currently, the compilation of a class definition will detect file modification in order to re-run the compilation, see:

https://github.com/CuyZ/Valinor/blob/dd4624c5e0376cf1c590117dcad10c659a614701/src/Definition/Repository/Cache/Compiler/ClassDefinitionCompiler.php#L67-L82

While this is often needed during a user's development phase (because the value-objects used for the mapping are always changing), this is useless in a production environment, where the PHP files should not change (or during a deployment phase, but the caches should then be cleared). 

The issue is that `filemtime` function is quite heavy and resource-hungry, where this library tries to be as optimized as possible.

## Proposal

The [`MapperBuilder`](https://github.com/CuyZ/Valinor/blob/dd4624c5e0376cf1c590117dcad10c659a614701/src/MapperBuilder.php) should give access to a new method that would allow to "activate" development mode — this setting would then be used by `ClassDefinitionCompiler` to handle, or not, the file modification feature.

**Comment by shulard** [comment]
Hello !

I'm not sure that having this kind of development mode is a nice addition. Can't we have a Decorator to encapsulate the specific development mode feature ? This way we can use composition to define what's the most relevant implementation for the current context. Combined with a dependency injection container this will be completely transparent for the end user ^^.

What do you think ?

**Comment by romm** [comment]
Hi, that's a very good proposal indeed; at some point during the library development I even implemented a service provider, as defined in the excellent [container-interop/service-provider](https://github.com/container-interop/service-provider) repository (that has its chances to become a PSR someday) — but I removed it, see below why.

The main problem I can see with making the container open to the world is that it would bring more maintenance because of poorly designed external implementations. Reducing the public API (also as described in #4) sounds like a good idea to me — I'd rather take time to analyze and work on a good submitted issue/pull-request that is strictly bound to the library codebase, than having to understand why an external implementation of some service crashes everything. To be honest, I'm not really used to such a level of contributions burden on any of the repositories I maintain, I've just been reading feedback of other people (cc @Ocramius 😛) and it inspired me to take this kind of decisions.

But that doesn't mean that we can't use the strength of the decorator pattern internally to encapsulate the "development mode" feature — still we have to find how the user would activate it; I'll wait to have your feedback on these statements to start digging into it. 🙂

**Comment by shulard** [comment]
I'm totally with you about the container. A library like this can give objects and public API that will be used to give a result. When I mentioned the container, it's more on the end user side. For example with Symfony, we can imagine a bridge package that handles all the wiring with the framework internal details (development mode for example). For sure this bridge will not land in the current implementation.

However relying on a decorator allows to configure the behaviour using composition then it allows an easy integration with containers.

Today you have you internal container and the « only » user configuration is the `Settings` object. To get started quickly, adding a setting that will be interpreted inside the container seems to be a good way to let the user configure it. Another solution will be to allow adding custom service definition inside your internal container but this seems too internal to be updated from the outside.

I'll try to use this package in a Symfony project soon, if it's relevant I can work on the bridge ^^

So let's move forward about this development mode. The mainenance burden must not be handle by only one person. For the moment, we have the easy task to add comments and give feedbacks ^^

If you are ok with the `Settings` way, I can work on a PR ^^

**Comment by romm** [comment]
> Today you have you internal container and the « only » user configuration is the Settings object. To get started quickly, adding a setting that will be interpreted inside the container seems to be a good way to let the user configure it. Another solution will be to allow adding custom service definition inside your internal container but this seems too internal to be updated from the outside.

I'm not sure I understood correctly your point, but the user *can* actually customize settings, by configuring the `MapperBuilder` (which is the only entry point into the library right now). See for instance:

https://github.com/CuyZ/Valinor/blob/dd4624c5e0376cf1c590117dcad10c659a614701/src/MapperBuilder.php#L30-L36

> So let's move forward about this development mode. The mainenance burden must not be handle by only one person. For the moment, we have the easy task to add comments and give feedbacks ^^
> 
> If you are ok with the Settings way, I can work on a PR ^^

Any contribution is more than welcome! We can keep the discussion about the implementation details here in the meantime. My first thoughts was to do something like this:

```php
final class MapperBuilder
{
    private Settings $settings;

    public function withDevelopmentMode(bool $flag): self
    {
        $clone = clone $this;
        $clone->settings->developmentMode = $flag;

        return $clone;
    }

    // …
}

(new MapperBuilder())
    ->withDevelopmentMode(env(APP_ENV) === 'dev')
    ->mapper()
    ->map(...);
```

WDYT? Maybe I missed something in your explanation. 😅

> Is performance really a problem right now?

Not really, it's just something that I've been thinking about for some time, I didn't want to forget it. 🙂

I'm mostly

>  trying out the library on small (isolated) examples right now, but shouldn't PSR-6 caching handle this?

Well, the caching system is already there, the issue is mostly about detecting when to invalidate the cache in case any change is made to the file that is being "compiled" into cache.

> In a production environment, I would expect the cache to have no TTL at all (for most use-cases), since types are fixed as per deployment, while in a development environment, I would expect no caching at all.
That's actually something I never considered (having no cache at all in dev-mode). But still, we have to know how the user tells the library it's in dev-mode.

**Comment by Ocramius** [comment]
> the issue is mostly about detecting when to invalidate the cache in case any change is made to the file that is being "compiled" into cache.

 * for `development`, use a null/void cache (always invalidate)
 * for `production`, always consider the cache as valid (sources should be considered immutable)

It's a fair assumption that covers `~99.99%` of codebases.

**Comment by romm** [comment]
>     * for `development`, use a null/void cache (always invalidate)
> 
>     * for `production`, always consider the cache as valid (sources should be considered immutable)
> 
> 
> It's a fair assumption that covers `~99.99%` of codebases.

Agreed, with a slight change that in dev-mode a memoization cache is used for the like of the current request. WDYT?

**Comment by Ocramius** [comment]
> memoization cache is used for the like of the current request.

Sensible: using an array-like in-memory PSR-6 cache suffices.

**Comment by romm** [comment]
Well that's what I meant by memoization 😬 I may be using the word badly though.

**Comment by romm** [comment]
Hi there, I just submitted #132. The discussions we had here inspired me on how to implement it, so I guess you might be interested. 🙂

**Comment by romm** [comment]
Closed with #132.
